Implementation of MK docs. (#33)
FAIRmat-NFDI/nyaml: docs/learn/workflow.md

## nyaml Workflow

> The `nyaml` implements a specific workflow for converting between YAML and XML formats. The `nyaml` converter checks for the correct file type and calls the appropriate converter. For an XML file, the `nxdl2yaml` converter parses the `XML` file, by means of [lxml](https://lxml.de/) python library, into an `XML` tree object. Adhering to the NXDL rules, the converter writes the application definition or the base class object to a `YAML` file that matches the `nyaml` syntax. If the input file is a `YAML` file, the `yaml2nxdl` converter collects the comments in a `Comments` object and parses the `YAML` file into a python `dictionary` object. Later, the application definitions or base classes will be converted into an `XML` file by combining the `Comments` and the python `dictionary` object.

Reviewer: ```suggestion
`nyaml` implements a specific workflow for converting between YAML and XML formats. The `nyaml` converter checks for the correct file type and calls the appropriate converter. For an XML file, the `nxdl2yaml` converter parses the `XML` file, by means of the [lxml](https://lxml.de/) Python library, into an `XML` tree object. Adhering to the NXDL rules, the converter writes the application definition or the base class object to a `YAML` file that matches the `nyaml` syntax. If the input file is a `YAML` file, the `yaml2nxdl` converter collects the comments in a `Comments` object and parses the `YAML` file into a python `dictionary` object. Afterwards, the application definitions or base classes is converted into an `XML` file by combining the `Comments` and the Python `dictionary` object.
```

Reviewer 2: "the application definitions or base classes" use singular from like for XML direction aforementioned

Author: Addressed!

> After the conversion from `XML` format to `YAML` format, the `nyaml` tool stores the `XML` contents at the end of the file under a hash created by Secure Hash Algorithm sha256 from `YAML` contents generated in this conversion process. This benefits on the backward compatibility of the `YAML` file, so that the `YAML` file can be converted back to `XML` file without any modification of structure like whitespaces or newlines.

Reviewer: ```suggestion
After the conversion from `XML` format to `YAML` format, the `nyaml` tool stores the `XML` contents at the end of the file under a hash created by Secure Hash Algorithm sha256 from `YAML` contents generated in this conversion process. This benefits the backward compatibility of the `YAML` file, so that the `YAML` file can be converted back to `XML` file without any modification of structure like whitespaces or newlines.
```

Reviewer 2: Rather a comment:

We should at least in one place though mention that this hashing is also used because of the fact that the round trip conversions do not contain order otherwise even if the tool were implemented more rigorously, the xsd tells clearly that complex types are sequences meaning the order of the XML attributes and concepts matters. For a normal set of concepts it does not. https://www.balisage.net/Proceedings/vol26/html/LaFontaine01/BalisageVol26-LaFontaine01.html and https://stackoverflow.com/questions/14917943/python-lxml-enforcing-a-specific-order-for-attributes where doing neither and caching is a practical workaround

https://lxml.de/FAQ.html
https://github.com/nexusformat/definitions/blob/8237a9c2c417d4d3d25f72c9e5e69a6707fca337/nxdl.xsd#L986

Author: Text for workflow explanation reviewed and updated!

Like every scientific software, the `nyaml` tool also follows a specific workflow.

```mermaid
graph TD;
  subgraph Start
    id1["Input File (YAML or XML)"]
  end
  subgraph Correct File Converter
    id2["nyaml2nxdl Converter"]
    id3["nxdl2nyaml Converter"]
  end
  subgraph nyaml2nxdl
    id4["Comment Collector"]
    id5["Python Dictionary Object"]
  end
  subgraph nxdl2nyaml
    id6["XML Object"]
  end
  subgraph Final result
  id7["Write XML File"]
  id8["Write YAML File"]
  end

  id1--> |YAML File|id2
  id1--> |XML File|id3
  id2-->id4
  id4-->id5
  id3-->id6
  id5-->id7
  id6-->id8
```

## Conversion from YAML to XML and vice versa

> Presented below is a concise and trimmed example of the `NXmpes` application definition (for discussion purpose) in `YAML` format, alongside its corresponding encoding into `XML` format, as illustrated below. Subsequently, the fundamental rules governing this conversion process are elucidated. For a comprehensive understanding of the basic structure of NXDL, readers are encouraged to explore the [NeXus Manual](https://manual.nexusformat.org/user_manual.html).

Reviewer: Can we link to NXmpes here?

Reviewer 2: The point about that NXmpes is picked here which is mentioned later that it is only an example should be pulled up here already

Author: Inlcuded in note!

!!! note

>     For indetailed explanation with examples please follow the [Tutorials for writing NeXus definition in YAML](../tutorials/tutorials.md).

Reviewer: ```suggestion
    For a detailed explanation with examples please follow the [tutorial for writing NeXus definition in YAML](../tutorials/tutorials.md).
```

Author: Refactored the description of the workflow.

!!! note

>     The `NXmpes` application is an old and trimmed version which does not serve any scientific purpose, but only is used to illustrate the results of the conversion process.

Reviewer: ```suggestion
    The application definition used here is an old and trimmed version of `NXmpes`  which does not serve any scientific purpose, but is only used to illustrate the results of the conversion process.
```

```suggestion
    The application definition used here is an old and trimmed version of `NXmpes`  which does not serve any scientific purpose, but is only used here to illustrate the results of the conversion process.
```

Author: Updated!

=== "YAML"
    ```yaml
    category: application
    type: group
    doc: |
      This is the most general application definition for multidimensional photoelectron spectroscopy.

      .. _ISO 18115-1:2023: https://www.iso.org/standard/74811.html
      .. _IUPAC Recommendations 2020: https://doi.org/10.1515/pac-2019-0404
    symbols:
      doc: |
        The symbols used in the schema to specify e.g. dimensions of arrays
      n_transmission_function: |
        Number of data points in the transmission function.
    NXmpes(NXobject):
      (NXentry):
        definition:
          \@version:
          enumeration: [NXmpes]
        title:
        start_time(NX_DATE_TIME):
          doc: |
            Datetime of the start of the measurement.
        end_time(NX_DATE_TIME):
          exists: recommended
          doc: |
            Datetime of the end of the measurement.
        (NXinstrument):
          doc:
          - |
            Description of the MPES spectrometer and its individual parts.
          - |
            xref:
              spec: ISO 18115-1:2023
              term: 12.58
              url: https://www.iso.org/obp/ui/en/#iso:std:iso:18115:-1:ed-3:v1:en:term:12.58

>           source_TYPE(NXsource):

Reviewer 2: nameType should be added even though this being an old example

Author: Added!
